README.md
# office-hours

Are you tired of your code working all day every day? Don't you feel bad that your code keeps working while you're off
relaxing and having fun after work?

Well now you can use the power of `office-hours` to only run your code within the working hours of the day!

> [!IMPORTANT]  
> At the time of writing, the office hours are determined from the **Local Time Zone**
> of the host machine where the code is running. I might consider updating this library
> to support other timezones if I **really** want to suffer :P

## Usage

1. Add this library to your project

    ```shell
    cargo add office-hours
    ```

2. Import and use the `OfficeHours` struct

    ```rust
    use office_hours::OfficeHours;

    fn main() {
        // 9am to 5pm are the default office hours
        let office_hours = OfficeHours::default();
        if office_hours.now() {
            println!("Blimey! Is it time for work already?");
        } else {
            println!("Phew, still on break!");
        }
    }
    ```

3. **(Optional)** Import and use the `office_hours!` macro

    ```rust
    use office_hours::office_hours;
   
    fn main() {
        office_hours!({
            println!("Blimey! Is it time for work already?");
        }) 
    }
    ```

   Further examples can be found in the [examples/](/examples) directory. (`cargo run --example default_hours`)

## Development

> [!NOTE]
> The Minimum Supported Rust Version is **1.60.0**

1. Clone repository

    ```shell
    git clone https://github.com/sgoudham/office-hours.git
    cd office-hours
    ```

2. Build

    ```shell
    cargo build --release
    ```

3. Test

    ```shell
    cargo test --verbose
    ```

## License

[MIT](./LICENSE)

CHANGELOG.md
# CHANGELOG

All notable changes to this project will be documented in this file.

## [v1.1.0](https://github.com/sgoudham/office-hours/releases/tag/v1.1.0) - 2023-10-20


### Features

- Use const fn where possible ([93a4b6f](https://github.com/sgoudham/office-hours/commit/93a4b6f))
- Tweak IntoIterator implementation ([6343256](https://github.com/sgoudham/office-hours/commit/6343256))
- Annotate pure functions with must_use ([39bcab0](https://github.com/sgoudham/office-hours/commit/39bcab0))


### Documentation

- Add backticks to inline code snippets ([e8ec301](https://github.com/sgoudham/office-hours/commit/e8ec301))



## [v1.0.0](https://github.com/sgoudham/office-hours/releases/tag/v1.0.0) - 2023-10-19


### Features

- Initial commit ([d27ef51](https://github.com/sgoudham/office-hours/commit/d27ef51))





tests/test.rs
#[cfg(test)]
mod macros {
    #[test]
    fn office_hours_should_compile() {
        let t = trybuild::TestCases::new();
        t.pass("tests/macro/success*.rs");
        t.compile_fail("tests/macro/failure*.rs");
    }
}

#[cfg(test)]
mod lib {
    mod now {
        use chrono::NaiveTime;

        use office_hours::{Clock, FromNaiveTime, OfficeHours};

        #[test]
        fn true_when_start_and_finish_are_equal() {
            let OfficeHours { start, finish } = OfficeHours::new(Clock::TwelvePm, Clock::TwelvePm);
            assert!(OfficeHours::now_from_time(
                &start,
                &finish,
                &NaiveTime::from_time(Clock::TwelvePm)
            ));
        }

        #[test]
        fn true_when_at_hour_boundary() {
            let OfficeHours { start, finish } = OfficeHours::default();
            let start_boundary = NaiveTime::from_slice_u32(&[9, 0, 0, 1]).unwrap();
            assert!(OfficeHours::now_from_time(&start, &finish, &start_boundary));
            let end_boundary = NaiveTime::from_slice_u32(&[16, 59, 59, 59]).unwrap();
            assert!(OfficeHours::now_from_time(&start, &finish, &end_boundary));
        }

        #[test]
        fn true_when_crossing_day_boundary() {
            let office_hours = OfficeHours::new(Clock::EightPm, Clock::FourAm);
            for i in office_hours.iter() {
                assert!(OfficeHours::now_from_time(
                    &office_hours.start,
                    &office_hours.finish,
                    &i
                ));
            }
        }

        #[test]
        fn default_happy_path() {
            let OfficeHours { start, finish } = OfficeHours::default();
            let current = NaiveTime::from_time(Clock::EightPm);
            assert!(!OfficeHours::now_from_time(&start, &finish, &current));

            // before office hours
            for i in 0..9 {
                assert!(!OfficeHours::now_from_time(
                    &start,
                    &finish,
                    &NaiveTime::from_time_u32(i as u32).unwrap()
                ))
            }
            // during office hours
            for i in 9..17 {
                assert!(OfficeHours::now_from_time(
                    &start,
                    &finish,
                    &NaiveTime::from_time_u32(i as u32).unwrap()
                ))
            }
            // after office hours
            for i in 17..23 {
                assert!(!OfficeHours::now_from_time(
                    &start,
                    &finish,
                    &NaiveTime::from_time_u32(i as u32).unwrap()
                ))
            }
        }
    }
}


tests/macro/success_closure.rs
use office_hours::office_hours;

fn main() {
    office_hours!(Clock::TwelveAm, Clock::ElevenPm, |start, end| {
        println!(
            "Man, I just love when this code works only within my macro hours of {:?} -> {:?}",
            start, end
        )
    });
    office_hours!(Clock::TwelveAm, Clock::ElevenPm, |start, _| {
        println!(
            "Man, I just love when this code works only within my macro hours of {:?} -> Unknown",
            start
        )
    });
    office_hours!(Clock::TwelveAm, Clock::ElevenPm, |_, end| {
        println!(
            "Man, I just love when this code works only within my macro hours of Unknown -> {:?}",
            end
        )
    });
}

tests/macro/success_default.rs
use office_hours::office_hours;

fn main() {
    office_hours!({ println!("I love my macro hours from 9am to 5pm") });
}

src/macro.rs
/// Creates an [`OfficeHours`] that takes in a list of statements to execute
/// when the current time is between office hours.
///
/// - Only execute code between the default hours of 9am and 5pm.
///
/// ```
/// use office_hours::office_hours;
/// office_hours!({ println!("Between 9am and 5pm") });
/// ```
///
/// - Only execute code between the custom hours of 5pm and 10pm.
///
/// ```
/// use office_hours::office_hours;
/// office_hours!(Clock::FivePm, Clock::TenPm, {
///     println!("Between 5pm and 10pm")
/// });
/// ```
/// [`OfficeHours`]: crate::OfficeHours
#[macro_export]
macro_rules! office_hours {
    ({ $($code:stmt)* }) => {{
        use office_hours::Clock;
        office_hours!(Clock::NineAm, Clock::FivePm, { $($code)* })
    }};
    ($start:expr, $finish:expr, { $($code:stmt)* }) => {{
        use office_hours::{Clock, OfficeHours};
        let office_hours = OfficeHours::new($start, $finish);
        if office_hours.now() {
            $($code)*
        }
    }};
    ($start_var:expr, $finish_var:expr, | $start:pat_param, $finish:pat_param | { $($code:stmt)* }) => {{
        use office_hours::{Clock, OfficeHours};
        let office_hours = OfficeHours::new($start_var, $finish_var);
        let $start = &office_hours.start;
        let $finish = &office_hours.finish;
        if office_hours.now() {
            $($code)*
        }
    }};
}


src/lib.rs
//! Give your code a break like any good employer!
//!
//! Are you tired of your code working all day every day? Don't you feel bad that your code keeps working while you're off
//! relaxing and having fun after work?
//!
//! Well now you can use the power of **office-hours** to only run your code within the working hours of the day!
//!
//! # Important
//!
//! At the time of writing, the office hours are determined from the **Local Time Zone**
//! of the host machine where the code is running. I might consider updating this library
//! to support other timezones if I **really** want to suffer :P
//!
//! # Quick Start
//!
//! ```
//! use chrono::Timelike;
//! use office_hours::{Clock, OfficeHours};
//!
//! // 9am to 5pm are the default office hours
//! let office_hours = OfficeHours::default();
//! if office_hours.now() {
//!     println!("Blimey! Is it time for work already?");
//! } else {
//!     println!("Phew, still on break!");
//! }
//!
//! // 12pm to 5pm
//! let afternoon = OfficeHours::new(Clock::TwelvePm, Clock::FivePm);
//! if afternoon.now() {
//!     println!("Blimey! Is it time for work already?");
//! } else {
//!     println!("Phew, still on break!");
//! }
//!
//! // We can also cross the day boundary - 10pm to 6am
//! let night_shift = OfficeHours::new(Clock::TenPm, Clock::SixAm);
//! if night_shift.now() {
//!     println!("Blimey! Is it time for work already?");
//! } else {
//!     println!("Phew, still on break!");
//! }
//!
//! // Iterate over office hours as chrono::NaiveTime
//! for time in office_hours.iter() {
//!     println!("Hour: {:?}", time.hour());
//!     println!("Minute: {:?}", time.minute());
//!     println!("Second: {:?}", time.second());
//! }
//!
//! // Iterate over office hours as u32
//! for hour in office_hours.hours_iter() {
//!     println!("Hour: {:?}", hour);
//! }
//! ```
#![warn(clippy::all, clippy::pedantic, clippy::nursery)]
use std::cmp::Ordering;

use chrono::{Duration, Local, NaiveTime, Timelike};

use thiserror::Error;
use OfficeHoursError::InvalidTimeSlice;

mod r#macro;

#[derive(Error, Debug)]
pub enum OfficeHoursError<'a> {
    #[error("Could not convert `{0:?}` to chrono::NaiveTime")]
    InvalidTimeSlice(&'a [u32]),
}

/// Simple enum to store all the valid hours in a day.
pub enum Clock {
    TwelveAm = 0,
    OneAm = 1,
    TwoAm = 2,
    ThreeAm = 3,
    FourAm = 4,
    FiveAm = 5,
    SixAm = 6,
    SevenAm = 7,
    EightAm = 8,
    NineAm = 9,
    TenAm = 10,
    ElevenAm = 11,
    TwelvePm = 12,
    OnePm = 13,
    TwoPm = 14,
    ThreePm = 15,
    FourPm = 16,
    FivePm = 17,
    SixPm = 18,
    SevenPm = 19,
    EightPm = 20,
    NinePm = 21,
    TenPm = 22,
    ElevenPm = 23,
}

/// Trait to implement helper functions in the style of [`From`].
pub trait FromNaiveTime {
    /// Makes a new [`NaiveTime`] out of a `&[u32]` slice.
    ///
    /// The slice supports a maximum length of 4 parts: `hour`,
    /// `minute`, `second`, and `millisecond`.
    ///
    /// If any part is missing, it is substituted with a `0`.
    ///
    /// If an empty slice is given, it will default
    /// to midnight.
    ///
    /// # Errors
    ///
    /// Returns [`Err`] if an invalid hour, minute, second,
    /// and/or millisecond.
    ///
    /// # Examples
    ///
    /// ```
    /// use chrono::NaiveTime;
    /// use office_hours::FromNaiveTime;
    ///
    /// let midnight = NaiveTime::from_slice_u32(&[]).unwrap();
    /// assert_eq!(midnight, NaiveTime::default());
    ///
    /// let nine = NaiveTime::from_slice_u32(&[9]).unwrap();
    /// assert_eq!(nine, NaiveTime::from_hms_milli_opt(9, 0, 0, 0).unwrap());
    ///
    /// let nine_thirty = NaiveTime::from_slice_u32(&[9, 30]).unwrap();
    /// assert_eq!(
    ///     nine_thirty,
    ///     NaiveTime::from_hms_milli_opt(9, 30, 0, 0).unwrap()
    /// );
    /// ```
    ///
    /// See [`NaiveTime::from_hms_milli_opt`] for further information.
    fn from_slice_u32(slice: &[u32]) -> Result<NaiveTime, OfficeHoursError>;

    /// Makes a new [`NaiveTime`] out of a [`Clock`].
    ///
    /// # Examples
    ///
    /// ```
    /// use chrono::NaiveTime;
    /// use office_hours::{Clock, FromNaiveTime};
    /// let twelve_pm = NaiveTime::from_time(Clock::TwelvePm);
    /// assert_eq!(
    ///     twelve_pm,
    ///     NaiveTime::from_hms_milli_opt(12, 0, 0, 0).unwrap()
    /// )
    /// ```
    fn from_time(time: Clock) -> NaiveTime;

    /// Makes a new [`NaiveTime`] from the hour given as a [`u32`].
    ///
    /// # Errors
    ///
    /// Returns [`None`] when given an invalid hour.
    ///
    /// # Examples
    ///
    /// ```
    /// use chrono::NaiveTime;
    /// use office_hours::FromNaiveTime;
    /// let twelve_pm = NaiveTime::from_time_u32(12).unwrap();
    /// assert_eq!(
    ///     twelve_pm,
    ///     NaiveTime::from_hms_milli_opt(12, 0, 0, 0).unwrap()
    /// )
    /// ```
    fn from_time_u32(hour: u32) -> Option<NaiveTime>;
}

impl FromNaiveTime for NaiveTime {
    fn from_slice_u32(slice: &[u32]) -> Result<NaiveTime, OfficeHoursError> {
        let time = Self::from_hms_milli_opt(
            *slice.first().unwrap_or(&0),
            *slice.get(1).unwrap_or(&0),
            *slice.get(2).unwrap_or(&0),
            *slice.get(3).unwrap_or(&0),
        )
        .ok_or(InvalidTimeSlice(slice))?;
        Ok(time)
    }

    fn from_time(hour: Clock) -> NaiveTime {
        // -- SAFETY --
        // Make sure all variants of the [Clock] enum continues
        // to hold hours of the day that valid (0 -> 23)
        unsafe { Self::from_time_u32(hour as u32).unwrap_unchecked() }
    }

    fn from_time_u32(hour: u32) -> Option<NaiveTime> {
        Self::from_hms_opt(hour, 0, 0)
    }
}

/// Simple struct to store the start and the finish
/// of the office day.
#[derive(Debug, Eq, PartialEq)]
pub struct OfficeHours {
    pub start: NaiveTime,
    pub finish: NaiveTime,
}

impl Default for OfficeHours {
    /// The default office hours are 9am to 5pm.
    ///
    /// # Examples
    ///
    /// ```
    /// use chrono::NaiveTime;
    /// use office_hours::{Clock, FromNaiveTime, OfficeHours};
    ///
    /// let office_hours = OfficeHours::default();
    /// assert_eq!(office_hours.start, NaiveTime::from_time(Clock::NineAm));
    /// assert_eq!(office_hours.finish, NaiveTime::from_time(Clock::FivePm));
    /// ```
    fn default() -> Self {
        Self {
            start: NaiveTime::from_time(Clock::NineAm),
            finish: NaiveTime::from_time(Clock::FivePm),
        }
    }
}

impl OfficeHours {
    /// Makes a new [`OfficeHours`] from the starting and finishing time.
    ///
    /// # Examples
    ///
    /// ```
    /// use office_hours::{Clock, OfficeHours};
    /// let morning = OfficeHours::new(Clock::NineAm, Clock::TwelvePm);
    /// let afternoon = OfficeHours::new(Clock::TwelvePm, Clock::FivePm);
    /// ```
    #[must_use]
    pub fn new(start: Clock, finish: Clock) -> Self {
        Self {
            start: NaiveTime::from_time(start),
            finish: NaiveTime::from_time(finish),
        }
    }

    /// Returns an iterator over the office hours.
    ///
    /// The iterator yields all [`NaiveTime`] instances from
    /// the starting hour to finishing hour.
    ///
    /// # Examples
    ///
    /// ```
    /// use office_hours::{Clock, OfficeHours};
    ///
    /// let office_hours = OfficeHours::default();
    /// for time in office_hours.iter() {
    ///     println!("Hours: {:?}", time);
    /// }
    /// ```
    #[must_use]
    pub const fn iter(&self) -> OfficeHoursIter {
        OfficeHoursIter {
            start: self.start,
            finish: self.finish,
        }
    }

    /// Collect the contents of [`Self::hours_iter`] into a [`Vec<u32>`].
    ///
    /// # Examples
    ///
    /// ```
    /// use office_hours::{Clock, OfficeHours};
    ///
    /// let office_hours = OfficeHours::default();
    /// assert_eq!(office_hours.hours(), vec![9, 10, 11, 12, 13, 14, 15, 16]);
    /// let night_shift = OfficeHours::new(Clock::TenPm, Clock::FiveAm);
    /// assert_eq!(night_shift.hours(), vec![22, 23, 0, 1, 2, 3, 4]);
    /// ```
    #[must_use]
    pub fn hours(&self) -> Vec<u32> {
        self.hours_iter().collect()
    }

    /// Returns an iterator over the office hours.
    ///
    /// The iterator yields all hours from
    /// the starting hour to finishing hour in its
    /// raw [`u32`] form.
    ///
    /// # Examples
    ///
    /// ```
    /// use office_hours::{Clock, OfficeHours};
    ///
    /// let office_hours = OfficeHours::default();
    /// for hour in office_hours.hours_iter() {
    ///     println!("u32: {:?}", hour);
    /// }
    /// ```
    pub fn hours_iter(&self) -> impl Iterator<Item = u32> {
        self.iter().map(|time| time.hour())
    }

    /// Checks if the current local time is within the office hours.
    ///
    /// # Examples
    ///
    /// ```
    /// use office_hours::OfficeHours;
    ///
    /// let office_hours = OfficeHours::default();
    /// if office_hours.now() {
    ///     println!("Blimey! Is it time for work already?")
    /// } else {
    ///     println!("Phew, still on break!")
    /// }
    /// ```
    #[must_use]
    pub fn now(&self) -> bool {
        Self::now_from_time(&self.start, &self.finish, &Local::now().time())
    }

    #[doc(hidden)]
    #[must_use]
    pub fn now_from_time(start: &NaiveTime, finish: &NaiveTime, now: &NaiveTime) -> bool {
        match start.cmp(finish) {
            Ordering::Equal => start == now,
            Ordering::Less => (start..finish).contains(&now),
            Ordering::Greater => now >= start || now < finish,
        }
    }
}

impl<'a> TryFrom<(&'a [u32], &'a [u32])> for OfficeHours {
    type Error = OfficeHoursError<'a>;

    /// Performs the conversion.
    ///
    /// # Examples
    ///
    /// ```
    /// use office_hours::OfficeHours;
    ///
    /// let start: &[u32] = &[9, 30];
    /// let finish: &[u32] = &[17, 30];
    /// let valid_office_hours = OfficeHours::try_from((start, finish));
    /// assert!(valid_office_hours.is_ok());
    ///
    /// let start_err: &[u32] = &[100000000];
    /// let invalid_office_hours = OfficeHours::try_from((start_err, finish));
    /// assert!(invalid_office_hours.is_err());
    /// ```
    fn try_from(office_hours: (&'a [u32], &'a [u32])) -> Result<Self, Self::Error> {
        let (start, finish) = office_hours;
        Ok(Self {
            start: NaiveTime::from_slice_u32(start)?,
            finish: NaiveTime::from_slice_u32(finish)?,
        })
    }
}

/// Iterator over [`OfficeHours`] that returns the hourly [`NaiveTime`].
pub struct OfficeHoursIter {
    start: NaiveTime,
    finish: NaiveTime,
}

impl Iterator for OfficeHoursIter {
    type Item = NaiveTime;

    fn next(&mut self) -> Option<NaiveTime> {
        if self.start == self.finish {
            None
        } else {
            let current_hour = NaiveTime::from_time_u32(self.start.hour()).unwrap();
            let is_before_midnight = self.start == NaiveTime::from_time(Clock::ElevenPm);
            self.start = if is_before_midnight {
                NaiveTime::default()
            } else {
                self.start + Duration::hours(1)
            };
            Some(current_hour)
        }
    }
}

impl IntoIterator for &OfficeHours {
    type Item = NaiveTime;
    type IntoIter = OfficeHoursIter;
    fn into_iter(self) -> Self::IntoIter {
        self.iter()
    }
}
